uplink/storage/streams/store.go

```diff
@@ -291,47 +291,48 @@ func (s *streamStore) upload(ctx context.Context, path Path, pathCipher storj.Ci
 			}
 		}
 
-		streamInfo, err := proto.Marshal(&pb.StreamInfo{
-			DeprecatedNumberOfSegments: currentSegment + 1,
-			SegmentsSize:               s.segmentSize,
-			LastSegmentSize:            sizeReader.Size(),
-			Metadata:                   metadata,
-		})
-		if err != nil {
-			return Meta{}, currentSegment, streamID, err
-		}
+		currentSegment++
+		lastSegmentSize = sizeReader.Size()
+		streamSize += lastSegmentSize
+	}
 
-		// encrypt metadata with the content encryption key and zero nonce
-		encryptedStreamInfo, err := encryption.Encrypt(streamInfo, s.cipher, &contentKey, &storj.Nonce{})
-		if err != nil {
-			return Meta{}, currentSegment, streamID, err
-		}
+	if eofReader.hasError() {
+		return Meta{}, currentSegment, streamID, eofReader.err
```

Comment:
Won't `currentSegment` in the case of an error here (and in other error conditions) be +1 compared to before this change? It looks like this value is used in the case of an error as `totalSegments` in a call to `cancelHandler`:

```
// CancelHandler handles clean up of segments on receiving CTRL+C
func (s *streamStore) cancelHandler(ctx context.Context, streamID storj.StreamID, totalSegments int64, path Path, pathCipher storj.CipherSuite) {
	defer mon.Task()(&ctx)(nil)

	for i := int64(0); i < totalSegments; i++ {
		err := s.segments.Delete(ctx, streamID, int32(i))
		if err != nil {
			zap.L().Warn("Failed deleting segment", zap.String("path", path.String()), zap.Int64("segmentIndex", i), zap.Error(err))
			continue
		}
	}
}
```

Assuming I'm not misunderstanding, does this mean that the old or new behavior is correct?

Reply:
It won't be for this case because this check was already outside of the loop (https://github.com/storj/storj/blob/master/uplink/storage/streams/store.go#L333), however, it will be for the others that they have been moved out of the loop (https://github.com/storj/storj/blob/master/uplink/storage/streams/store.go#L294L327).

So, it looks that we had the problem of asking for one more segment before and now we have in a few more cases.

There are 2 solutions.

1. Exclude the not found errors in `cancelHandler` and not showing a warning for them.
2. Align the indexes out of the loop correctly.

Solution 1 doesn't avoid to make extra request meanwhile the solution 2 does.

The solution 2 doesn't require that much work but considering that the bug already existed in some cases and I have a another PR (https://github.com/storj/storj/pull/3471) piling up on top of this, I would prefer to have the fix after the merge because I want to avoid conflicts as much as possible on in order of saving time.

By the way, good catch!

Reply:
I [created a ticket](https://storjlabs.atlassian.net/secure/RapidBoard.jspa?rapidView=12&modal=detail&selectedIssue=V3-3139&quickFilter=101) for this so it doesn't get forgotten but I'm not sure I understand the big picture enough to write a good description. @ifraixedes, would you mind improving the title/description as you see fit?

Reply:
sure, I'll take care of it. Thanks.
I didn't create it because I hope to make the change after this PR, but it's good to have it to remind me.

Reply:
PR: https://github.com/storj/storj/pull/3555